django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{params['page_size']}"
+        return cache_vary
+
+    def get_queryset(self, params) -> QuerySet[PackageListing]:
+        # Prep Q objects
+        filter_q = Q()
+        exclude_q = Q()
+
+        # Add majority of filters
+        if params.get("community_identifier", None):
+            filter_q.add(Q(community__identifier=params["community_identifier"]), Q.AND)
+
+        if params.get("namespace_identifier", None):
+            filter_q.add(
+                Q(package__namespace__name=params["namespace_identifier"]), Q.AND
+            )
+
+        if params.get("team_identifier", None):
+            filter_q.add(Q(package__owner__name=params["team_identifier"]), Q.AND)
+
+        if params.get("user_identifier", None):
+            filter_q.add(
+                Q(
+                    package__owner__members__user__username__icontains=params[
+                        "user_identifier"
+                    ]
+                ),
+                Q.AND,
+            )
+
+        if params.get("package_identifier", None):
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
+
+        if not params.get("deprecated", None):
+            exclude_q.add(Q(package__is_deprecated=True), Q.OR)
+
+        if not params.get("nsfw", None):
+            exclude_q.add(Q(has_nsfw_content=True), Q.OR)
+
+        if params.get("q", None):
+            icontains_query = Q()
+            parts = [x for x in params["q"].split(" ") if x]
+
+            for part in parts:
+                for field in (
+                    "package__name",
+                    "package__owner__name",
+                    "package__latest__description",
+                ):
+                    icontains_query &= ~Q(**{f"{field}__icontains": part})
+            exclude_q.add(icontains_query, Q.OR)
+
+        # Add category filters/excludes
+        included = []
+        excluded = []
+        if params.get("section", None):
+            try:
+                section = PackageListingSection.objects.prefetch_related(
+                    "require_categories", "exclude_categories"
+                ).get(slug=params["section"])
+            except PackageListingSection.DoesNotExist:
+                pass
+            else:
+                [included.append(s.slug) for s in section.require_categories.all()]
+                [excluded.append(s.slug) for s in section.exclude_categories.all()]
+
+        if params.get("included_categories", None):
+            included = list(chain(included, params["included_categories"]))
+
+        if params.get("excluded_categories", None):
+            excluded = list(chain(excluded, params["excluded_categories"]))
+
+        if len(included) > 0:
+            included_cats_q = Q()
+            for cat in sorted(set(included)):
+                included_cats_q.add(Q(categories__slug__icontains=cat), Q.OR)
```

Comment:
Use exact comparisons instead of `icontains`

Reply:
Sure, thought it would be a bit more client friendly to use `icontains`, since I didn't see a situation where the case would be a deciding factor. (with categories specifically)

Reply:
It would not, it would return unexpected results because it does partial matching instead of requiring a full match. Even if that wasn't a problem, doing an exact lookup is faster so should be preferred.

Reply:
Oh, I read "use `contains` instead of..." at first :D Sure I'll change that.